File: code/python/501.py
# https://leetcode.com/problems/find-mode-in-binary-search-tree/

class Solution:
    def findMode(self, root: Optional[TreeNode]) -> List[int]:
        def dfs(root):
            nonlocal maxGroup, currGroup, currVal, res
            if not root:
                return
            dfs(root.left)

            if root.val == currVal:
                currGroup += 1
            else:
                currGroup = 1
                currVal = root.val
            if currGroup > maxGroup:
                res = []
                maxGroup = currGroup
            if currGroup == maxGroup:
                res.append(root.val)

            dfs(root.right)

        maxGroup = currGroup = currVal = 0
        res = []
        dfs(root)
        return res

File: code/cpp/501.cpp
// https://leetcode.com/problems/find-mode-in-binary-search-tree/

class Solution {
public:
    vector<int> res;
    int maxGroup = 0, currGroup = 0, currVal = 0;

    void dfs(TreeNode* root) {
        if (!root)
            return;
        dfs(root->left);

        if (root->val == currVal)
            currGroup++;
        else {
            currGroup = 1;
            currVal = root->val;
        }
        if (currGroup > maxGroup) {
            res = {};
            maxGroup = currGroup;
        }
        if (currGroup == maxGroup)
            res.emplace_back(root->val);

        dfs(root->right);
    }

    vector<int> findMode(TreeNode* root) {
        dfs(root);
        return res;
    }
};
